### PR TITLE
[v7r1] singularityCE: added possibility to run without re-installing DIRAC inside the container

### PR DESCRIPTION
**Pull request by fstagni**
@sfayer I added a flag that can be used for running without re-installing DIRAC inside the container, but by default the previous behavior stays. To be honest I think that the new behavior (simply `source bashrc`) should *always* work iff Pilot3 is used. And it should work also for DIRAC extension, and for Pilot extensions (if the Pilot extension saves the environment in bashrc, e.g. see https://gitlab.cern.ch/lhcb-dirac/LHCbPilot/-/blob/master/LHCbPilot/LHCbPilotCommands.py#L160). In LHCb we'll certainly run with the new flag on (so, without DIRAC re-installations).

There's another thing: as described in https://its.cern.ch/jira/browse/LHCBDIRAC-863 (I doubt you can see this one... ):

> At the moment, we are not able to run a full DIRAC client in singularity because of the VOMS and X509 environment variables propagation.
> Currently: if set by the system, they are not set. Otherwise, we set them to CVMFS.
> When running singularity these variables may be already set as `/etc/grid-security/*`. However inside the container, this location is overwritten, and the voms directory is empty.
> Possible solution:
> - Remount each env variable at the same time when starting the container
> - Mount /etc/grid-security as /etc/grid-security always (can do because RPM standard)
> - mount /cvmfs/.../grid-security as /etc/grid-security
> - Force a change of env variable inside the container to always use CVMFS

I made one test using the `ContainerBind` option and got it working (2nd option above), but Worker Nodes of certain sites might be re-setting these variables to other locations so I am not convinced that what we have is 100% sufficient.

Anyway, it needs more testing (e.g. with https://github.com/DIRACGrid/Pilot/pull/117) but I didn't have time to set it up and right now the DIRAC certification machine is broken.

BEGINRELEASENOTES

*Resources
NEW: SingularityCE: singularityCE: added possibility to run without re-installing DIRAC inside the container

*docs:
CHANGE: improved doc for SingularityCE options

ENDRELEASENOTES

### CODE REVIEW

File: Resources/Computing/SingularityComputingElement.py [hunk 1]
```diff
@@ -342,14 +377,18 @@ def submitJob(self, executableFile, proxy=None, **kwargs):
...
+    cmd.extend(["--ipc", "--pid"])  # run container in new IPC and PID namespaces
+    cmd.extend(["--workdir", baseDir])  # working directory to be used for /tmp, /var/tmp and $HOME
     if withCVMFS:
-      cmd.extend(["-B", "/cvmfs"])
+      cmd.extend(["--bind", "/cvmfs"])
     if 'ContainerBind' in self.ceParameters:
       bindPaths = self.ceParameters['ContainerBind'].split(',')
       for bindPath in bindPaths:
-        cmd.extend(["-B", bindPath.strip()])
+        if len(bindPath.split('=')) == 1:
```
Comment (sfayer) [inline Resources/Computing/SingularityComputingElement.py:388]: I don't understand the significance of the '=' character in this block... The example in the documentation doesn't contain an equals, so I'm wondering if it's meant to be ':::' instead?

**Comment by sfayer** [comment]
@fstagni I think even with pilot3, the `source bashrc` would fail in the vanilla DIRAC case because it still does a local install that wouldn't be visible in the container (only LHCb have the magic LHCbDIRACInstall module that symlinks to CVMFS).

The X509 certificates dir is a bit of a difficult one to fix. Sites do tend to have /etc/grid-security, but certainly at Imperial it's practically empty (the same as inside the container), I imagine some other sites are the same. My initial thought on this were to bind in X509_CERT_DIR if it's outside of CVMFS and then copy the environment variable though. In my opinion that has too many special cases if the site has done something special.
I think the only safe option here is to get it from CVMFS by default (which is the correct choice for the CernVM container). Perhaps this will have to be another CE option that defaults to the CVMFS path?

There is one more general case that I've just thought of while reading through this, which you'll probably run into as you do more wider testing: the singularity binary itself on CVMFS relying on user namespaces (the user namespaces are recommended by WLCG everywhere now to avoid setuid binaries) . This is a case of adding a CVMFS search path onto the end of places to look in the `__hasSingularity` function. If the file /proc/sys/user/max_user_namespaces contains a number bigger than 100, the singularity arguments should have the --userns option added. I can open a new issue about this if you'd prefer to deal with it separately.

Regards,
Simon

File: Resources/Computing/SingularityComputingElement.py (hunk 1 above)
Comment (fstagni) [reply Resources/Computing/SingularityComputingElement.py:390]: Yes of course, thanks for noticing. Fixing it.

**Comment by fstagni** [comment]
> @fstagni I think even with pilot3, the `source bashrc` would fail in the vanilla DIRAC case because it still does a local install that wouldn't be visible in the container (only LHCb have the magic LHCbDIRACInstall module that symlinks to CVMFS).

Yes, that's true. But we should start moving for vanilla DIRAC for CVMFS-based installation a-la-LHCb, still managed by dirac-install. It's not there yet but it would work.

> The X509 certificates dir is a bit of a difficult one to fix. Sites do tend to have /etc/grid-security, but certainly at Imperial it's practically empty (the same as inside the container), I imagine some other sites are the same. My initial thought on this were to bind in X509_CERT_DIR if it's outside of CVMFS and then copy the environment variable though. In my opinion that has too many special cases if the site has done something special.
> I think the only safe option here is to get it from CVMFS by default (which is the correct choice for the CernVM container). Perhaps this will have to be another CE option that defaults to the CVMFS path?

Maybe, but:
- the CVMFS location would be VO-dependent
- some sites DON'T have CVMFS (mostly HPCs) and for them it's yet another solution

> There is one more general case that I've just thought of while reading through this, which you'll probably run into as you do more wider testing: the singularity binary itself on CVMFS relying on user namespaces (the user namespaces are recommended by WLCG everywhere now to avoid setuid binaries) . This is a case of adding a CVMFS search path onto the end of places to look in the `__hasSingularity` function. If the file /proc/sys/user/max_user_namespaces contains a number bigger than 100, the singularity arguments should have the --userns option added. I can open a new issue about this if you'd prefer to deal with it separately.

Yes please. But TBH I would be even happier if you directly create the Pull Request for it.

File: docs/source/AdministratorGuide/Configuration/ConfReference/Resources/Computing/index.rst [hunk 2]
```diff
@@ -123,15 +123,24 @@ ARC CE Parameters
...
-+------------------------+--------+----------------------------------------------------------+---------------------------------------------+
-| ContainerRoot          | The root image location for the container to use.                 |  /cvmfs/cernvm-prod.cern.ch/cvm3            |
-+------------------------+--------+----------------------------------------------------------+---------------------------------------------+
-| ContainerExtraOpts     | Extra options for dirac-install within the container.             |  -u 'http://other.host/instdir' -g 'v13r0'  |
-+------------------------+--------+----------------------------------------------------------+---------------------------------------------+
-| KeepWorkArea           | If set to True container work area won't be deleted at end of job |  True (Default: False)                      |
-+------------------------+--------+----------------------------------------------------------+---------------------------------------------+
++------------------------ +-------------------------------------------------------------------+---------------------------------------------  +
```
Comment (andresailer) [inline docs/source/AdministratorGuide/Configuration/ConfReference/Resources/Computing/index.rst:126]:
```suggestion
+------------------------ +-------------------------------------------------------------------+-----------------------------------------------+
```
I don't think the spaces are allowed there? Applies for all these lines.

File: Resources/Computing/SingularityComputingElement.py (hunk 1 above)
Comment (sfayer) [reply Resources/Computing/SingularityComputingElement.py]: The second index on line 393 should probably be [1] too...

**Comment by sfayer** [comment]
For the X509_CERT_DIR, how about this simple logic (written in some python-like pseudo code, I'm sure you can work out what I mean):
```
cert_dir = realpath(env.get("X509_CERT_DIR", "/etc/grid-security/certificates"))
# Probably need to catch the exception if target of the above doesn't actually exist.
Bind cert_dir to /etc/grid-security/certificates inside the container.
```
(And then similar for X509_VOMS_DIR). The certificates should be working outside the container, so the target should always be good and the /etc/grid-security directory tree exists inside the CernVM container so it won't create any overlayfs issues. The realpath function is just in-case a site has a symlink in X509_CERT_DIR pointing to somewhere else, I doubt there are (m)any sites that do that.

Sure, I'll try to create a PR for the userns stuff next week.

File: Resources/Computing/SingularityComputingElement.py [hunk 3]
```diff
@@ -342,14 +379,18 @@ def submitJob(self, executableFile, proxy=None, **kwargs):
...
-      cmd.extend(["-B", "/cvmfs"])
+      cmd.extend(["--bind", "/cvmfs"])
     if 'ContainerBind' in self.ceParameters:
       bindPaths = self.ceParameters['ContainerBind'].split(',')
       for bindPath in bindPaths:
-        cmd.extend(["-B", bindPath.strip()])
+        if len(bindPath.split(':::')) == 1:
+          cmd.extend(["--bind", bindPath.split(':::')[0].strip()])
+        elif len(bindPath.split(':::')) == 2:
+          cmd.extend(["--bind", bindPath.split(':::')[0].strip(), bindPath.split(':::')[0].strip()])
```
Comment (atsareg) [inline Resources/Computing/SingularityComputingElement.py:393]: Did yo mean:

        if len(bindPath.split(':::')) == 1:
          cmd.extend(["--bind", bindPath.strip()])
        elif len(bindPath.split(':::')) == 2:
          cmd.extend(["--bind", bindPath.split(':::')[0].strip(), bindPath.split(':::')[1].strip()])

Comment (fstagni) [reply Resources/Computing/SingularityComputingElement.py]: Yes, that's better. Applied.

**Comment by fstagni** [comment]
I think this PR is good to me merged (after review). The issue raised with certificates dir will be handled later, for the moment I would just to go on and test what's here.

**Review by atsareg — APPROVED** [review]

**Pull request merged**